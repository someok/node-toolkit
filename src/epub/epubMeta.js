const _ = require('lodash');
const {loadToc} = require('./toc');
const {getTitle, getAuthor} = require('../utils/titleUtils');
const {success, failure} = require('../utils/result');
const {loadMetadataYaml} = require('../metadata/yaml');
const {init} = require('../metadata/metadata');
const Meta = require('../metadata/Meta');

/**
 * 读取给定目录下的 meta 信息，如果不存在 __t2e.data 及下面的 yaml，则根据当前文件名自动生成。
 *
 * @param folder 目录
 * @return {Result} 返回 {@link Result}
 */
exports.readMetadata = function(folder) {
    const tocNodes = loadToc(folder);
    if (_.isEmpty(tocNodes)) {
        return failure('此文件内不存在任何 txt 文本文件');
    }

    let metaJson = {
        title: getTitle(folder),
        author: getAuthor(folder),
        description: null,
    };
    // 如果 metaJson 信息不存在则创建，否则忽略
    const initResult = init(folder, Meta.fromJson(metaJson), false);
    if (!initResult.success) {
        return initResult;
    }

    const meta = loadMetadataYaml(folder);

    return success({
        meta,
        tocNodes,
    });
};
