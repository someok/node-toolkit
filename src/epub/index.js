const {logError} = require('../utils/logUtils');
const {createTempFolder} = require('../utils/fileUtils');
const {genEpubFile} = require('../utils/zipUtils');
const {generate} = require('./generate');
const {readMetadata} = require('./epubMeta');

function genEpub(txtDir, epubFile) {
    const result = readMetadata(txtDir);
    if (!result.success) {
        logError(result.message);
        return;
    }

    const {meta, tocNodes} = result.data;

    const tmpDir = createTempFolder();
    console.log(tmpDir);
    generate(tmpDir, meta, tocNodes);

    genEpubFile(tmpDir, epubFile);
}

module.exports = genEpub;
