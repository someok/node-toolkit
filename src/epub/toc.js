const path = require('path');
const fs = require('fs');
const klawSync = require('klaw-sync');
const _ = require('lodash');
const pingyin = require('pinyinlite');

const mdListParser = require('../utils/marked/list2JsonParser');
const {logError} = require('../utils/logUtils');
const TxtNode = require('../utils/TxtNode');
const {TOC_FILE, METADATA_FOLDER} = require('../context');

function loadToc(folder) {
    const tocPath = path.resolve(folder, METADATA_FOLDER, TOC_FILE);
    if (fs.existsSync(tocPath)) {
        const mdContent = fs.readFileSync(tocPath);
        const mdTocNodes = mdListParser(mdContent.toString());

        return mdTocNodes;
    }

    // 不存在 toc.md 文件，则直接读取 txt 文件名作为目录
    return loadAllTxtFileNames(folder, true);
}

/**
 * 文件过滤，支持 txt 扩展名。
 *
 * todo: 将来可以考虑增加 md 扩展名。
 *
 * @param item 文件路径
 * @return {boolean} 是否有效文本文件
 */
function txtFilter(item) {
    return ['.txt'].includes(path.extname(item.path).toLowerCase());
}

/**
 * 返回给定文件夹下所有 txt 文本文件名，并按照拼音排序。
 *
 * 如果文件名是 01_file_real_name.txt 格式，则会返回 file_real_name，前置的 01 用于排序。
 *
 * @param folder txt 文件夹
 * @param pinyinSort 是否按照拼音排序
 * @return {Array} txt 文件名构成的数组，每个条目包括：{rawTitle, title, ext, pinyin}
 */
function loadAllTxtFileNames(folder, pinyinSort = false) {
    const files = klawSync(folder, {
        nodir: true,
        filter: txtFilter,
    });

    if (_.isEmpty(files)) {
        return null;
    }

    const fileArr = files.map(file => {
        // 如果标题前面有数字（格式为：123__xxx），则去掉下划线及前置数字，只保留后面内容
        const re = /^(\d*_{2})?(.+)/g;

        const filePath = file.path;
        const ext = path.extname(filePath);
        const rawTitle = path.basename(filePath, ext).trim();
        const match = re.exec(rawTitle);
        let title = rawTitle;
        if (match) {
            title = match[2];
        }

        return {
            rawTitle,
            title,
            ext,
            path: filePath,
        };
    });

    if (pinyinSort) {
        fileArr.sort((f1, f2) => {
            const f1Pinyin = pingyin(f1.rawTitle).join('');
            const f2Pinyin = pingyin(f2.rawTitle).join('');
            return f1Pinyin.localeCompare(f2Pinyin);
        });
    }

    return fileArr;
}

module.exports = {
    loadAllTxtFileNames,
    loadToc,
};
