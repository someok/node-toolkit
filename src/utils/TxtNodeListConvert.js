const _ = require('lodash');
const TxtNode = require('./TxtNode');

function getChapterId(index) {
    return _.padStart(index, 4, '0');
}

function getHref(index) {
    return `chapter-${getChapterId(index)}.xhtml`;
}

function toNavMap(txtNodes) {
    let ncx = '<navMap>\n';

    let index = 0;

    function travelNode(nodes) {
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            const title = node.title;
            const {children} = node;

            const hasChildren = Array.isArray(children) && children.length > 0;
            ncx += '<navPoint>\n';

            if (node.path) {
                ncx += `<navLabel><text>${title}</text></navLabel>\n`;
                ncx += `<content src="${getHref(index++)}"/>\n`;
            } else {
                ncx += `<navLabel><text>${title}</text></navLabel>\n`;
            }

            if (hasChildren) {
                travelNode(children);
            }
            ncx += '</navPoint>';
        }
    }

    travelNode(txtNodes);

    ncx += '</navMap>';

    return ncx;
}

function toHtmlOrderList(txtNodes) {
    let ol = '<ol>\n';

    let index = 0;

    function travelNode(nodes) {
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            const title = node.title;
            const {children} = node;

            const hasChildren = Array.isArray(children) && children.length > 0;
            ol += '<li>\n';

            if (node.path) {
                ol += `<a href="${getHref(index++)}">${title}</a>\n`;
            } else {
                ol += `<span>${title}</span>`;
            }

            if (hasChildren) {
                ol += '<ol>\n';
                travelNode(children);
                ol += '</ol>\n';
            }
            ol += '</li>';
        }
    }

    travelNode(txtNodes);

    ol += '</ol>';
    return ol;
}

function toChapters(txtNodes) {
    let index = 0;
    const chapterIds = [];
    TxtNode.travelTxtNodeTree(txtNodes, node => {
        if (node.path) {
            chapterIds.push({
                id: getChapterId(index++),
                title: node.title,
                path: node.path,
            });
        }
    });

    return chapterIds;
}

module.exports = {
    toNavMap,
    toHtmlOrderList,
    toChapters,
};
