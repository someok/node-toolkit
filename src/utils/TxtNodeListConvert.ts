import TxtNode from './TxtNode';

export function toNavMap(txtNodes: TxtNode[]): string {
    let ncx = '<navMap>\n';

    function travelNode(nodes: TxtNode[]): void {
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            const title = node.title;
            const {children} = node;

            const hasChildren = Array.isArray(children) && children.length > 0;
            ncx += '<navPoint>\n';

            if (node.chapterId) {
                ncx += `<navLabel><text>${title}</text></navLabel>\n`;
                // noinspection HtmlDeprecatedTag
                ncx += `<content src="${node.getHref()}"/>\n`;
            } else {
                ncx += `<navLabel><text>${title}</text></navLabel>\n`;
            }

            if (hasChildren) {
                travelNode(children);
            }
            ncx += '</navPoint>';
        }
    }

    travelNode(txtNodes);

    ncx += '</navMap>';

    return ncx;
}

export function toHtmlOrderList(txtNodes: TxtNode[]): string {
    let ol = '<ol>\n';

    function travelNode(nodes: TxtNode[]): void {
        for (let i = 0; i < nodes.length; i++) {
            const node = nodes[i];
            const title = node.title;
            const {children} = node;

            const hasChildren = Array.isArray(children) && children.length > 0;
            ol += '<li>\n';

            if (node.chapterId) {
                ol += `<a href="${node.getHref()}">${title}</a>\n`;
            } else {
                ol += `<span>${title}</span>`;
            }

            if (hasChildren) {
                ol += '<ol>\n';
                travelNode(children);
                ol += '</ol>\n';
            }
            ol += '</li>';
        }
    }

    travelNode(txtNodes);

    ol += '</ol>';
    return ol;
}

interface Chapter {
    id: string;
    title?: string;
    path?: string;
}

export function toChapters(txtNodes: TxtNode[]): Chapter[] {
    const chapterIds: Chapter[] = [];
    TxtNode.travelTxtNodeTree(
        txtNodes,
        (node: TxtNode): void => {
            if (node.chapterId) {
                chapterIds.push({
                    id: node.getPadChapterId(),
                    title: node.title,
                    path: node.path,
                });
            }
        }
    );

    return chapterIds;
}
