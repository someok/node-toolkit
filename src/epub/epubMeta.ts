import {getAuthor, getTitle} from '../utils/titleUtils';
import Result, {success} from '../utils/result';

import {loadToc} from './toc';
import {loadMetadataYaml} from '../metadata/yaml';
import metadataInit from '../metadata/metadata';
import Meta from '../metadata/Meta';

/**
 * 读取给定目录下的 meta 信息，如果不存在 __t2e.data 及下面的 yaml，则根据当前文件名自动生成。
 *
 * @param folder 目录
 * @return {Result} 返回 {@link Result}
 */
export function readMetadata(folder: string): Result {
    const tocResult = loadToc(folder);
    if (!tocResult.success) {
        return tocResult;
    }

    let metaJson = {
        title: getTitle(folder),
        author: getAuthor(folder),
        description: null,
    };
    // 如果 metaJson 信息不存在则创建，否则忽略
    const initResult = metadataInit(folder, Meta.fromJson(metaJson), false);
    if (!initResult.success) {
        return initResult;
    }

    const meta = loadMetadataYaml(folder);

    return success({
        meta,
        tocNodes: tocResult.data,
    });
};
