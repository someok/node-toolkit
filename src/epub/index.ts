import fse from 'fs-extra';
import path from 'path';
import _ from 'lodash';

import {logInfo} from '@someok/node-utils/lib/logUtils';
import {createTempFolder, existDir} from '@someok/node-utils/lib/fileUtils';
import {zipDir} from '@someok/node-utils/lib/zipUtils';

import {subdirs} from '../utils/fileUtils';
import {generate} from './generate';
import {readMetadata} from './epubMeta';
import Meta from '../metadata/Meta';

/**
 * 将给定 txt 文件夹转换为 epub 文件。
 *
 * @param txtDir txt 文件路径
 * @param epubPath 以 .epub 为扩展名表示文件路径，否则为所在目录，如果是后者，则用 meta 中定义的名称命名
 */
export function genTxtDir2Epub(txtDir: string, epubPath: string): Promise<Meta> {
    if (!txtDir) {
        return Promise.reject(new Error('必须明确 txt 文件夹'));
    }
    if (!existDir(txtDir)) {
        return Promise.reject(new Error(`[${txtDir}] 不是合法文件夹`));
    }
    if (!epubPath) {
        return Promise.reject(new Error('生成 epub 的路径必须明确'));
    }

    const tmpDir = createTempFolder();
    return new Promise<Meta>(function(resolve, reject): void {
        readMetadata(txtDir)
            .then(
                ({meta, tocNodes}): Promise<void> => {
                    generate(tmpDir, meta, tocNodes);

                    let epubFile: string;
                    if (epubPath.toLowerCase().endsWith('.epub')) {
                        epubFile = epubPath;
                    } else {
                        epubFile = path.join(epubPath, meta.epubTitle());
                    }

                    return zipDir(tmpDir, epubFile)
                        .then(
                            (): void => {
                                logInfo(`epub 生成生成：[${epubFile}]`);
                                resolve(meta);
                            }
                        )
                        .catch(
                            (err): void => {
                                reject(err);
                            }
                        );
                }
            )
            .catch(
                (err): void => {
                    reject(err);
                }
            )
            .finally(
                (): void => {
                    // 删除临时文件夹
                    fse.removeSync(tmpDir);
                }
            );
    });
}

export function genAllTxtDir2Epub(txtDir: string, epubPath: string): Promise<boolean> {
    const dirs = subdirs(txtDir);

    if (_.isEmpty(dirs)) {
        return Promise.reject(new Error(`[${txtDir}] 中没有子文件夹`));
    }

    const errMsg: string[] = [];

    // 顺序执行 Promise
    const seqPromise = dirs.reduce(function(promiseChain, dir): Promise<boolean> {
        return promiseChain.then(function(): Promise<boolean> {
            return genTxtDir2Epub(dir.path, epubPath)
                .then((): boolean => true)
                .catch(
                    (err): boolean => {
                        errMsg.push(err.message);
                        return false;
                    }
                );
        });
    }, Promise.resolve(true));

    return seqPromise.then(
        (): boolean => {
            if (!_.isEmpty(errMsg)) {
                throw new Error(errMsg.join('\n'));
            }

            return true;
        }
    );
}
