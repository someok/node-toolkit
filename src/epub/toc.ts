import path from 'path';
import fs from 'fs';
import klawSync from 'klaw-sync';
import _ from 'lodash';
import pingyin from 'pinyinlite';

import mdListParser from '../utils/marked/list2JsonParser';
import Result, {failure, success} from '../utils/result';
import {existPath, PathMode} from '../utils/fileUtils';
import TxtNode from '../utils/TxtNode';
import {METADATA_FOLDER, TOC_FILE} from '../context';

/**
 * 从给定文件夹中读取目录列表。
 *
 * 首先读取 toc.md 文件，如果此文件不存在，则读取 txt 文件名作为目录。
 *
 * @param folder 文件夹
 * @return {Result} 目录列表或错误信息
 */
export function loadToc(folder: string): Result {
    const tocPath = path.resolve(folder, METADATA_FOLDER, TOC_FILE);

    if (fs.existsSync(tocPath)) {
        return loadMdContentAsToc(folder, tocPath);
    }

    // 不存在 toc.md 文件，则直接读取 txt 文件名作为目录
    return loadTxtNamesAsToc(folder, true);
}

/**
 * 在 md 的树形列表中递归。
 *
 * @param {Array} nodes {@link TxtNode} 节点
 * @param {function} fn 遍历节点的时候执行的 callback 方法
 * @param level 树形层级，默认从 0 开始
 */
export function travelTxtNodeTree(nodes: TxtNode[], fn: Function, level: number = 0) {
    for (let i = 0; i < nodes.length; i++) {
        const node = nodes[i];
        const {children} = node;

        const hasChildren = Array.isArray(children) && children.length > 0;
        fn && fn(node, hasChildren, level);

        if (hasChildren) {
            travelTxtNodeTree(children, fn, level + 1);
        }
    }
}

/**
 * 在 md 的树形列表中递归，并设置每个节点的 path 属性，并判断此路径是否实际存在。
 *
 * @param folder 文件夹
 * @param {Array} nodes {@link TxtNode} 节点
 * @param {Array} notExistPath 不存在的节点路径
 */
function travelTree(folder: string, nodes: TxtNode[], notExistPath: string[]) {
    TxtNode.travelTxtNodeTree(nodes, (node: TxtNode, hasChildren: boolean) => {
        const {rawTitle} = node;
        const nodePath = rawTitle ? path.join(folder, rawTitle) : folder;
        const mode = existPath(nodePath);

        // 非分支节点允许非物理路径存在，也就是说父节点可以只作为标题存在
        // 此时 node.path 为空
        if (hasChildren) {
            if (mode === PathMode.IS_FILE) {
                node.path = nodePath;
            }
        } else {
            if (mode === PathMode.IS_FILE) {
                node.path = nodePath;
            } else {
                rawTitle && notExistPath.push(rawTitle);
            }
        }
    });
}

/**
 * 遍历 md toc 节点，并检测对应文件是否存在，不存在给出提示，存在则将 path 属性置入 Node
 *
 * @param folder txt 所在文件夹
 * @param mdFile toc.md 文件，markdown 列表格式
 * @return {Result} {@link TxtNode} 数组
 */
export function loadMdContentAsToc(folder: string, mdFile: string): Result {
    const mdName = path.basename(mdFile);
    let mdContent;
    try {
        mdContent = fs.readFileSync(mdFile);
    } catch (e) {
        return failure(`[${mdName}] 文件读取出现错误：${e.message}`);
    }

    mdContent = mdContent.toString().trim();
    if (!mdContent.trim()) return failure(`[${mdName}] 文件内容尚未定义`);

    const mdTocNodes: TxtNode[] | null = mdListParser(mdContent);
    if (!mdTocNodes || _.isEmpty(mdTocNodes)) return failure(`[${mdName}] 文件内容尚未定义`);

    // 递归遍历
    const notExistPath: string[] = [];
    travelTree(folder, mdTocNodes, notExistPath);

    if (!_.isEmpty(notExistPath)) {
        let msg = `[${mdName}] 中如下节点并不实际存在:\n`;
        notExistPath.forEach(dir => {
            msg += `    - ${dir}\n`;
        });
        return failure(msg);
    }

    return success(mdTocNodes);
}

/**
 * 文件过滤，支持 txt 扩展名。
 *
 * todo: 将来可以考虑增加 md 扩展名。
 *
 * @param item 文件路径
 * @return {boolean} 是否有效文本文件
 */
function txtFilter(item: klawSync.Item) {
    return ['.txt'].includes(path.extname(item.path).toLowerCase());
}

/**
 * 返回给定文件夹下所有 txt 文本文件名作为目录，并按照拼音排序。
 *
 * 如果文件名是 01_file_real_name.txt 格式，则会返回 file_real_name，前置的 01 用于排序。
 *
 * @param folder txt 文件夹
 * @param pinyinSort 是否按照拼音排序
 * @return {Result} txt 文件名构成的数组，条目为 {@link TxtNode}
 */
export function loadTxtNamesAsToc(folder: string, pinyinSort: boolean = false): Result {
    const files = klawSync(folder, {
        nodir: true,
        filter: txtFilter,
    });

    if (_.isEmpty(files)) {
        return failure('未发现任何文本文件');
    }

    const fileArr = files.map(file => {
        // 如果标题前面有数字（格式为：123__xxx），则去掉下划线及前置数字，只保留后面内容
        const re = /^(\d*_{2})?(.+)/g;

        const filePath = file.path;
        const ext = path.extname(filePath);
        const rawTitle = path.basename(filePath).trim();
        let title = path.basename(filePath, ext).trim();
        const match = re.exec(title);
        if (match) {
            // todo: 增加 html escape 处理
            title = match[2];
        }

        return new TxtNode(title, undefined, undefined, rawTitle, ext, filePath);
    });

    if (pinyinSort) {
        fileArr.sort((f1, f2) => {
            const f1Pinyin = pingyin(f1.rawTitle).join('');
            const f2Pinyin = pingyin(f2.rawTitle).join('');
            return f1Pinyin.localeCompare(f2Pinyin);
        });
    }

    return success(fileArr);
}
