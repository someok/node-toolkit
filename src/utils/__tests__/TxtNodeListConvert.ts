import path from 'path';

import {toHtmlOrderList, toNavMap} from '../TxtNodeListConvert';
import {loadToc} from '../../epub/toc';

test('toNavMap', () => {
    const tocDir = path.resolve(__dirname, 'toc1');
    const result = loadToc(tocDir);
    const tocNodes = result.data;
    // console.log(tocNodes);
    const str = toNavMap(tocNodes);
    // console.log(str);
    expect(str).toMatchSnapshot();
});

test('toHtmlOrderList', () => {
    const tocDir = path.resolve(__dirname, 'toc1');
    const result = loadToc(tocDir);
    const tocNodes = result.data;
    // console.log(tocNodes);
    const str = toHtmlOrderList(tocNodes);
    // console.log(str);
    expect(str).toMatchSnapshot();
});
