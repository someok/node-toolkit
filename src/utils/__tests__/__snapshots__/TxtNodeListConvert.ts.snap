// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHtmlOrderList 1`] = `
"<ol>
<li>
<a href=\\"chapter-0001.xhtml\\">toc0</a>
<ol>
<li>
<a href=\\"chapter-0002.xhtml\\">toc01</a>
<ol>
<li>
<a href=\\"chapter-0003.xhtml\\">toc011</a>
</li></ol>
</li><li>
<a href=\\"chapter-0004.xhtml\\">toc02</a>
</li><li>
<a href=\\"chapter-0005.xhtml\\">toc03</a>
<ol>
<li>
<a href=\\"chapter-0006.xhtml\\">toc031</a>
</li></ol>
</li></ol>
</li><li>
<a href=\\"chapter-0007.xhtml\\">toc1</a>
</li><li>
<a href=\\"chapter-0008.xhtml\\">toc2</a>
<ol>
<li>
<a href=\\"chapter-0008.xhtml\\">toc21</a>
</li><li>
<a href=\\"chapter-0009.xhtml\\">toc22</a>
<ol>
<li>
<a href=\\"chapter-0010.xhtml\\">toc221</a>
</li></ol>
</li></ol>
</li></ol>"
`;

exports[`toNavMap 1`] = `
"<navMap>
<navPoint>
<navLabel><text>toc0</text></navLabel>
<content src=\\"chapter-0001.xhtml\\"/>
<navPoint>
<navLabel><text>toc01</text></navLabel>
<content src=\\"chapter-0002.xhtml\\"/>
<navPoint>
<navLabel><text>toc011</text></navLabel>
<content src=\\"chapter-0003.xhtml\\"/>
</navPoint></navPoint><navPoint>
<navLabel><text>toc02</text></navLabel>
<content src=\\"chapter-0004.xhtml\\"/>
</navPoint><navPoint>
<navLabel><text>toc03</text></navLabel>
<content src=\\"chapter-0005.xhtml\\"/>
<navPoint>
<navLabel><text>toc031</text></navLabel>
<content src=\\"chapter-0006.xhtml\\"/>
</navPoint></navPoint></navPoint><navPoint>
<navLabel><text>toc1</text></navLabel>
<content src=\\"chapter-0007.xhtml\\"/>
</navPoint><navPoint>
<navLabel><text>toc2</text></navLabel>
<content src=\\"chapter-0008.xhtml\\"/>
<navPoint>
<navLabel><text>toc21</text></navLabel>
<content src=\\"chapter-0008.xhtml\\"/>
</navPoint><navPoint>
<navLabel><text>toc22</text></navLabel>
<content src=\\"chapter-0009.xhtml\\"/>
<navPoint>
<navLabel><text>toc221</text></navLabel>
<content src=\\"chapter-0010.xhtml\\"/>
</navPoint></navPoint></navPoint></navMap>"
`;
